pkg/controller/sparkapplication/submission.go

```diff
@@ -367,3 +380,76 @@ func addExecutorConfOptions(app *v1beta2.SparkApplication, submissionID string)
 
 	return executorConfOptions, nil
 }
+
+// addLocalDirConfOptions excludes local dir volumes, update SparkApplication and returns local dir config options
+func addLocalDirConfOptions(app *v1beta2.SparkApplication) ([]string, error) {
+	var localDirConfOptions []string
+
+	sparkLocalVolumes := map[string]v1.Volume{}
+	var mutateVolumes []v1.Volume
+
+	for _, volume := range app.Spec.Volumes {
+		if strings.HasPrefix(volume.Name, config.SparkLocalDirVolumePrefix) {
+			sparkLocalVolumes[volume.Name] = volume
+		} else {
+			mutateVolumes = append(mutateVolumes, volume)
+		}
+	}
+	app.Spec.Volumes = mutateVolumes
+
+	if app.Spec.Driver.VolumeMounts != nil {
+		var mutateMountVolumes []v1.VolumeMount
+		for _, volumeMount := range app.Spec.Driver.VolumeMounts {
+			if volume, ok := sparkLocalVolumes[volumeMount.Name]; ok {
+				driverOptions := buildLocalVolumeOptions(config.SparkDriverVolumesPrefix, volume, volumeMount)
+				for _, option := range driverOptions {
+					localDirConfOptions = append(localDirConfOptions, option)
+				}
+			} else {
+				mutateMountVolumes = append(mutateMountVolumes, volumeMount)
+			}
+		}
+
+		app.Spec.Driver.VolumeMounts = mutateMountVolumes
+	}
+
+	if app.Spec.Executor.VolumeMounts != nil {
```

Comment:
This code block looks almost identical to the one above. I think we should create a function for it.

Reply:
Sure. Let me do a refactor.